kb/how-to/add-test-generation.md

```diff
@@ -0,0 +1,478 @@
+            )
+        # Call description() in the parent class
+        return super().description()
+
+    def result(self) -> str:
+        # Return as a quoted hexadecimal string, without leading 0x
+        return "\"{:x}\"".format(self.int_a + self.int_b)
+
+    @classmethod
+    def generate_function_tests(cls) -> Iterator[test_case.TestCase]:
+        # Create a list of input values
+        input_values = ["", "0", "1", "123"]
+        # combinations_with_replacement generates a list of all unique
+        # combinations, including where inputs are the same i.e. ("0", "0")
+        for a_value, b_value in itertools.combinations_with_replacement(
+            input_values, 2
+        ):
+                # Initialize a class instance with these values
+                class_instance = cls(a_value, b_value)
+                # Yield the TestCase object from this instance
+                yield class_instance.create_test_case()
+
+
+if __name__ == '__main__':
+    # Pass command line arguments and a description for the script
+    test_generation.main(sys.argv[1:], "Generate bignum tests.")
+```
+
+Running this script will create 10 test cases in `test_suite_mpi.ex.generated.data`.
+The tests can then be generated and ran together with other suites by running `make test`.
+To run only these tests:
+
+```sh
+$ python tests/scripts/generate_bignum_ex_tests.py
+$ make tests/test_suite_mpi.ex.generated
+$ cd tests
+$ ./test_suite_mpi.ex.generated
+```
+
+## Extending existing test classes
+
+When adding test case generation for functions, there may be previously implemented
+classes which generate tests for similar functions.
+For example, the test function `mpi_add_abs()` is very similar to `mpi_add_mpi()`, but
+returns an absolute result.
+Test structure, input format, and useful test cases may also be common across test
+functions.
+For example, a variety of test functions for binary operations use two inputs A and B,
+and one output, and can be implemented similarly.
+
+### Variants
+
+For variants of a test function, inheritance can be used to reduce repetition of code.
+For example, to implement the test function `mpi_add_abs()`, `BignumAddAbs` can inherit
+from the `BignumAdd` class.
+The necessary changes for implementing `BignumAddAbs` are then reduced:
+ - `count` should be reset to 0 for the class.
+ - `test_function` should be set to `mpi_add_mpi_abs`.
+ - `test_name` should be set.
+ - `int` forms of the inputs (`int_a` and `int_b`) should be made absolute.
+
+For example, possible implementation of this class:
+
+```python
+class BignumAddAbsExample(BignumAddExample):
+    """Tests for absolute variant of bignum addition."""
+    count = 0
+    test_function = "mpi_add_abs"
+    test_name = "MPI add (abs)"
+
+# Either override __init__()
+    def __init__(self, val_a, val_b) -> None:
+        super().__init__(val_a, val_b)
+        self.int_a = abs(self.int_a)
+        self.int_b = abs(self.int_b)
```

Comment:
I guess this works due to the way `BignumAdd` is structured, because `__init__` stores the representation of the A and B arguments in fields that aren't `int_a` and `int_b` but does not itself calculate the representation of the expected-result argument. But if the reader doesn't know that, they'll be completely lost. Besides, it's confusing to show the same piece of code with three parts where you're only supposed to take parts 1+2 or 1+3. I suggest not showing this approach at all, or in a different example.